sys/kern/signal.c

```diff
@@ -129,14 +131,47 @@ int do_sigprocmask(int how, const sigset_t *set, sigset_t *oset) {
     return EINVAL;
   }
 
-  if (sig_pending(td) < NSIG) {
+  if (sig_pending(td)) {
     WITH_SPIN_LOCK (&td->td_spin)
       td->td_flags |= TDF_NEEDSIGCHK;
   }
 
   return 0;
 }
 
+int do_sigsuspend(proc_t *p, const sigset_t *mask) {
+  thread_t *td = thread_self();
+  assert(td->td_proc == p);
+
+  WITH_PROC_LOCK(p) {
+    assert((td->td_pflags & TDP_OLDMASK) == 0);
+
+    td->td_oldmask = td->td_sigmask;
+    td->td_pflags |= TDP_OLDMASK;
+    do_sigprocmask(SIG_SETMASK, mask, NULL);
+
+    /*
+     * We want the sleep to be interrupted only if there's an actual signal
+     * to be handled, but _sleepq_wait() returns immediately if TDF_NEEDSIGCHK
+     * is set, so we clear the flag here if there are no real pending signals.
+     */
+    WITH_SPIN_LOCK (&td->td_spin) {
+      if (sig_pending(td))
+        td->td_flags |= TDF_NEEDSIGCHK;
+      else
+        td->td_flags &= ~TDF_NEEDSIGCHK;
+    }
+  }
+
+  int error;
+  while ((error = sleepq_wait_intr(td, "sigsuspend()")) == 0)
```

Comment:
When does spurious wakeup happen?

Reply:
It's not supposed to happen, but It's what FreeBSD does [here](http://bxr.su/FreeBSD/sys/kern/kern_sig.c#1550). Now that I think about it, an assertion would be better.
Maybe something along the lines of [`kpause`](http://bxr.su/NetBSD/sys/kern/kern_synch.c#230) would be handy here.

Reply:
I'm not sure if it would be better, but I don't mind it if you prefer it that way. I have made the corresponding change.